### ISSUE TITLE
Drop leading articles when sorting feeds

### ISSUE THREAD

**Issue by woorst** [newsboat/newsboat#310]
`feed-sort-order title` does a naive and awkward (IMO) alphabetical sort. There should be an option like `ignore-leading-articles yes`

So instead of this:
```
An Awful Podcast
Awesome Podcast
Better Podcast
The Best Podcast
```

We'd have this:
```
Awesome Podcast
An Awful Podcast
The Best Podcast
Better Podcast
```

**Comment by der-lyse** [comment]
In order to make this work we need a list of articles for each language and also need to know the actual language the feed (title) is in. The latter one could be problematic as this information might not be present in the feed. And I wouldn't wonder if there are languages out there with terms which are not only articles but have other meanings as well, doing more harm than good. So a naive implementation probably won't do the job properly.

In the meantime as a workaround you might want to sort the feeds manually in your _urls_ file. Does this do the trick?

**Comment by woorst** [comment]
> In the meantime as a workaround you might want to sort the feeds manually in your urls file. Does this do the trick?

No, I already have a way I like to keep my *urls* file organized.

> In order to make this work we need a list of articles for each language and also need to know the actual language the feed (title) is in.

I suppose considering in/definite articles in all languages at the same time would be complicated. Why not have the configuration option `ignore-leading-articles` available to ignore a/an/the but have it set to *no* by default?

**Comment by Minoru** [comment]
> In order to make this work we need [...] to know the actual language the feed (title) is in. The latter one could be problematic as this information might not be present in the feed.

On a list of 342 feeds (some personal, some news sites, some random junk), I see 24 Atom feeds that contain at least one `xml:lang` attribute, and 104 RSS feeds with `<language>` element. In both cases I also checked that the value of the attribute or tag is not empty. So:
- two thirds of the feeds don't contain language information;
- out of those that do, RSS is five times more popular than Atom, which is important since RSS only allows to specify language per-feed, not per-item.

So that's a problem. On the other hand, any given user probably speaks only a couple of languages, so it's feasible to just let them list the ones they know, and use that as a default when `<language>` and `xml:lang` are absent.

We can even assume the first language from locale, which should Just Work for a lot of people.


As for workarounds: one can try to use a `filter:` to strip the articles from the titles before feeds are fed into Newsboat.

> Why not have the configuration option ignore-leading-articles available to ignore a/an/the but have it set to no by default?

Once implemented, it's impossible to remove, nor change the semantics of, the configuration variable. So let's not rush, and think this through ;)

In this particular case, I'm very worried by the fact that the variable has a broad name but will only work for English.

Also, Newsboat has a ton of options already, so if we can somehow avoid adding another one, that's a success in my book.

By the way, how should this work in the presence of prefixes of some kind. Imagine a series of posts with titles like: "Books I Liked: The Song of Ice and Fire", "Books I Liked: Solaris". What order would  you expect these to be in?

Off the top of my head, I can't recall any systems that do this kind of sorting (even though I'm certain I've seen that sorting before). Can you, @woorst? Maybe we can learn something useful from existing implementations by e.g. looking at the settings they provide, or how they handle this in UI (e.g. do they let the user to switch between lexicographic and this language-specific sorting).

Oh, oh and last but not least: does this sorting even has a name?

**Comment by Minoru** [comment]
Wrote a wall of text but forgot to say the most important thing: @woorst, thanks for bringing this up! It's an interesting idea, and even though I don't know what to do about it yet, I appreciate you opening an issue to discuss that.

**Comment by woorst** [comment]
> Off the top of my head, I can't recall any systems that do this kind of sorting (even though I'm certain I've seen that sorting before).

It's pretty much the standard in media players, for example [ncmpcpp](https://github.com/arybczak/ncmpcpp/blob/81cb7a4f85287ce2671beca9023048f4de41ddf0/doc/config#L434) and iTunes. Check out this wikipedia [page](https://en.wikipedia.org/wiki/List_of_books_banned_by_governments) for another example of this sorting style in action. It is not uncommon.

> By the way, how should this work in the presence of prefixes of some kind.

I think only *leading* in/definite articles should be ignored. I think things would become too complicated if you tried to account for some arbitrary prefix.

So in your example I'd expect the following order because 'S' is before 'T':
- Books I Liked: **S**olaris
- Books I Liked: **T**he Song of Ice and Fire

But if there were no prefix I'd expect the same order but this time because 'L' comes before 'N':
- So**l**aris
- The So**n**g of Ice and Fire

> Oh, oh and last but not least: does this sorting even has a name?

I don't believe so. I've been searching and although I can find many examples of people sorting strings like this I can't find a name for this convention.

One last thing. I know this sorting style is common in english but I don't know how common it is in other languages.

**Comment by Minoru** [comment]
Yeah, I came up with a real bad example to show the difference :)

Russian and Ukrainian don't have articles, so no "article-less sorting" there. I know French has articles, but don't know if they treat them in any special way during sorting.

**Comment by der-lyse** [comment]
[Wikipedia](https://en.wikipedia.org/wiki/Alphabetical_order#The_and_other_common_words) mentions it, but doesn't know a name for that sorting either.

In German this kind of sorting exists in dictionaries and encyclopedias, but I can't think of other common use cases. A mate suggested to simply call it "encyclopedia order". ;-)
